File: backend/app/schemas/audit.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field


class AuditLogResponse(BaseModel):
    id: uuid.UUID
    organization_id: uuid.UUID | None
    user_id: uuid.UUID | None
    entity_type: str
    entity_id: uuid.UUID | None
    action: str
    before_data: dict | None
    after_data: dict | None
    metadata: dict | None = Field(alias="metadata_")
    ip_address: str | None
    request_id: str | None
    created_at: datetime

    model_config = {"from_attributes": True, "populate_by_name": True}


class AuditLogListResponse(BaseModel):
    items: list[AuditLogResponse]
    total: int
    page: int
    page_size: int

File: frontend/src/api/audit.ts
import { apiFetch } from "./client";

export type AuditLogItem = {
  id: string;
  organization_id: string | null;
  user_id: string | null;
  entity_type: string;
  entity_id: string | null;
  action: string;
  before_data: Record<string, unknown> | null;
  after_data: Record<string, unknown> | null;
  metadata: Record<string, unknown> | null;
  ip_address: string | null;
  request_id: string | null;
  created_at: string;
};

export type AuditLogList = {
  items: AuditLogItem[];
  total: number;
  page: number;
  page_size: number;
};

export type AuditFilters = {
  entity_type?: string;
  entity_id?: string;
  user_id?: string;
  action?: string;
  date_from?: string;
  date_to?: string;
  page?: number;
  page_size?: number;
};

export async function fetchAuditLogs(filters: AuditFilters = {}) {
  const query = new URLSearchParams();
  Object.entries(filters).forEach(([key, value]) => {
    if (value !== undefined && value !== "") query.set(key, String(value));
  });
  return apiFetch<AuditLogList>(`/api/v1/audit-logs?${query.toString()}`);
}
